### PR TITLE
Log job submit API deduplication

### PR DESCRIPTION
**Pull request by severinson**
This PR adds job submission deduplication to the log submit API. It also adds a key-value store backed by Postgres, which is used for the deduplication.

### CODE REVIEW

File: internal/armada/server.go [hunk 1]
```diff
@@ -196,6 +198,27 @@ func Serve(config *configuration.ArmadaConfig, healthChecks *health.MultiChecker
 		}
 		submitServerToRegister = pulsarSubmitServer
 
+		// If postgres details were provided, enable deduplication.
+		log.Infof("DedupTable: %s\n", config.Pulsar.DedupTable)
+		log.Infof("Postgres: %+v\n", config.Pulsar.Postgres)
```
Comment (d80tb7) [inline internal/armada/server.go:203]: be careful here else you might end up logging postgres passwords.

File: internal/armada/server/submit_to_log.go [hunk 2]
```diff
@@ -158,21 +170,74 @@ func (srv *PulsarSubmitServer) SubmitJobs(ctx context.Context, req *api.JobSubmi
...
 
+// removeDuplicateSubmissions filters out job submit request items with previously seen client ids.
+// Client ids are namespaced by queue.
+func (srv *PulsarSubmitServer) removeDuplicateSubmissions(ctx context.Context, queue string, items []*api.JobSubmitRequestItem) ([]*api.JobSubmitRequestItem, error) {
+
+	// For storage efficiency, we store hashes instead of user-provided strings.
+	// For computational efficiency, we create a lookup table to avoid computing the same hash twice.
+	// Client ids are namespaced by queue. Hence, we hash the client id together with the queue.
+	jobSubmitRequestItems := make([]*api.JobSubmitRequestItem, 0, len(items))
+	combinedHashData := make([]byte, 40)
```
Comment (d80tb7) [inline internal/armada/server/submit_to_log.go:198]: is 40 bytes enough here?

File: internal/armada/server.go (hunk 1 above)
Comment (severinson) [reply internal/armada/server.go:203]: You're right; removed.

File: internal/armada/server/submit_to_log.go (hunk 2 above)
Comment (severinson) [reply internal/armada/server/submit_to_log.go:198]: Yes, the compiler guarantees that each hash is exactly 20 bytes.

Comment (d80tb7) [reply internal/armada/server/submit_to_log.go:198]: ALso if you're just worried about size in memory then string interning would probably be better.  If you're worried about the size on disk in the db then hashes work

File: internal/pgkeyvalue/pgkeyvalue.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+// AddKey is equivalent to Add(ctx, key, nil).
+func (c *PGKeyValueStore) AddKey(ctx context.Context, key string) (bool, error) {
+	return c.Add(ctx, key, nil)
+}
+
+func (c *PGKeyValueStore) createTable(ctx context.Context) error {
+	var pgErr *pgconn.PgError
+	_, err := c.db.Exec(ctx, fmt.Sprintf("create table %s (key text primary key, value bytea, inserted timestamp not null);", c.tableName))
+	if errors.As(err, &pgErr) && pgErr.Code == "42P07" { // Relation already exists (i.e., someone else just created it), which is fine.
```
Comment (d80tb7) [inline internal/pgkeyvalue/pgkeyvalue.go:81]: I think pgx has some constants for errors

File: internal/pgkeyvalue/pgkeyvalue.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+		})
+	}
+	if tableName == "" {
+		return nil, errors.WithStack(&armadaerrors.ErrInvalidArgument{
+			Name:    "TableName",
+			Value:   tableName,
+			Message: "TableName must be non-empty",
+		})
+	}
+	cache, err := simplelru.NewLRU(100, nil)
```
[the remark this answers is not in the document]
Comment (severinson) [reply internal/pgkeyvalue/pgkeyvalue.go]: Agreed; changed to be a parameter and set to one million.

File: internal/pgkeyvalue/pgkeyvalue.go (hunk 3 above)
Comment (severinson) [reply internal/pgkeyvalue/pgkeyvalue.go]: Nice. I've updated the code to use them.

File: e2e/pulsar_test/pulsar_test.go [hunk 5]
```diff
@@ -118,6 +118,85 @@ func TestSubmitJobs(t *testing.T) {
...
+			originalJobIds[i] = res.JobResponseItems[i].GetJobId()
+		}
+
+		numEventsExpected := numJobs * 6
+		_, err = receiveJobSetSequences(ctx, consumer, armadaQueueName, req.JobSetId, numEventsExpected, 10*time.Second)
+		if err != nil {
+			return err
+		}
+
+		// The second time, all jobs should be rejected.
```
Comment (d80tb7) [inline e2e/pulsar_test/pulsar_test.go:149]: nit- they don't get rejected just the deduplicated ids are returned

**Review by d80tb7 — APPROVED** [review]

**Review by d80tb7 — APPROVED** [review]

**Pull request merged**